### PR TITLE
rgw/admin: add suport for link/unlink bucket

### PR DESCRIPTION
**Pull request by elrondwong**
This change implements link/unlink bucket APIs of
rgw/admin.

Related to #645 

Signed-off-by: wanggangfeng <gfengwong@gmail.com>

## Checklist
- [x] Added tests for features and functional changes
- [x] Public functions and types are documented
- [x] Standard formatting is applied to Go code
- [x] Is this a new API? Is this new API marked PREVIEW?

### CODE REVIEW

**Comment by ansiwen** [comment]
Thank you! Will look at it soon. 🙂

**Comment by ansiwen** [comment]
I have some early feedback already. First, the linter check failed, have a look. You need to add comments to all the newly added exported methods and types. Please also add the ` PREVIEW` and run `make api-update` as described [here](https://github.com/ceph/go-ceph/blob/master/docs/development.md#documentation-conventions). Thanks!

**Comment by elrondwong** [comment]
Thank you for your guidance!
I have pushed two commits for above question.

**Comment by elrondwong** [comment]
@ansiwen

File: docs/api-status.json [hunk 1]
```diff
@@ -1834,6 +1834,14 @@
...
         "name": "API.RemoveUserCap",
         "comment": "RemoveUserCap removes the capabilities from a user.\nOn Success, it returns the updated list of UserCaps for the user.\n"
+      },
+      {
+        "name": "API.LinkBucket",
+        "comment": "LinkBucket - https://docs.ceph.com/en/latest/radosgw/adminops/#link-bucket\n"
+      },
+      {
+        "name": "API.UnlinkBucket",
+        "comment": "UnlinkBucket - https://docs.ceph.com/en/latest/radosgw/adminops/#unlink-bucket\n"
```
Comment (ansiwen) [inline docs/api-status.json:1844]: This is still in the wrong section (should be in "preview", not in "stable"). Please remove the file and recreate with `make api-update`.

**Review by ansiwen — CHANGES_REQUESTED** [review]

File: rgw/admin/link_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+			UID:      "test",
+		})
+		assert.NoError(t, err)
+
+		bucket, err = co.GetBucketInfo(context.Background(), Bucket{Bucket: suite.bucketTestName})
+		assert.NoError(t, err)
+		assert.Equal(t, bucket.Owner, "test")
+	})
+
+	suite.T().Run("get user quota but no user is specified", func(t *testing.T) {
```
Comment (thotz) [inline rgw/admin/link_test.go:61]: I guess quota feature unrelated to this use case

**Comment by elrondwong** [comment]
I've re-run make api-update, and resolved review comments,re-run test, The version docs should be now ok, PTAL.

File: rgw/admin/link_test.go [hunk 3]
```diff
@@ -16,10 +16,10 @@ func (suite *RadosGWTestSuite) TestLink() {
 	co, err := New(suite.endpoint, suite.accessKey, suite.secretKey, newDebugHTTPClient(http.DefaultClient))
 	assert.NoError(suite.T(), err)
 
-	suite.T().Run("create test user", func(t *testing.T) {
-		user, err := co.CreateUser(context.Background(), User{ID: "test", DisplayName: "test-user", Email: "test@example.com"})
+	suite.T().Run("create test user1", func(t *testing.T) {
+		user, err := co.CreateUser(context.Background(), User{ID: "test1", DisplayName: "test-user1", Email: "test1@example.com"})
```
Comment (thotz) [inline rgw/admin/link_test.go:20]: IMO better to keep ID as test-user1/2 that test1/2 it helps in readablity

**Review by thotz — COMMENTED** [review]
Overall LGTM, but minor nits :)

File: rgw/admin/link.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	}
+	if link.Bucket == "" {
+		return errMissingBucket
+	}
+	_, err := api.call(ctx, http.MethodPost, "/bucket", valueToURLParams(link))
+	return err
+}
+
+// LinkBucket will link a bucket to a specified user
+// unlinking the bucket from any previous user
```
Comment (thotz) [inline rgw/admin/link.go:33]: @elrondwong : Sorry for the dump question, I don't understand completely about this comment, if you link with New user, then previous user will automatically unlinked??

Comment (elrondwong) [reply rgw/admin/link.go:33]: yes, I have try, but the unlink api makes no sense.

Comment (elrondwong) [reply rgw/admin/link.go:33]: [official docs](https://docs.ceph.com/en/pacific/radosgw/adminops/#unlink-bucket) show that.

Comment (thotz) [reply rgw/admin/link.go:33]: > yes, I have tried, but the unlink api makes no sense.

AFAIR user cannot be removed until all the buckets attached to it are removed, so if u want to keep data then just unlink the user for the time being and link the bucket with a new user whenever it creates. Now a day

**Review by thotz — APPROVED** [review]

File: rgw/admin/link.go (hunk 4 above)
Comment (elrondwong) [reply rgw/admin/link.go:33]: Yes! I just tested it and it's still today.
Thank you both for great reviews  and for guidance!!! 🥳

**Comment by elrondwong** [comment]
PTAL @leseb when you have some time,

**Review by leseb — APPROVED** [review]

**Comment by elrondwong** [comment]
@ansiwen All checks have passed, please merge.

**Comment by ansiwen** [comment]
@mergifyio squash

**Review by ansiwen — APPROVED** [review]

**Comment by elrondwong** [comment]
close #645

**Pull request merged**